core/src/main/java/org/apache/iceberg/io/RollingContentFileWriter.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.iceberg.ContentFile;
+import org.apache.iceberg.ContentFileWriter;
+import org.apache.iceberg.ContentFileWriterFactory;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Metrics;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.util.Tasks;
+
+public class RollingContentFileWriter<ContentFileT, T> implements Closeable {
+
+  private static final int ROWS_DIVISOR = 1000;
+  private final PartitionKey partitionKey;
+  private final FileFormat format;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+  private final ContentFileWriterFactory<ContentFileT, T> writerFactory;
+
+  private final WriterResult.Builder resultBuilder;
+
+  private EncryptedOutputFile currentFile = null;
+  private ContentFileWriter<ContentFileT, T> currentFileWriter = null;
+  private long currentRows = 0;
+
+  public RollingContentFileWriter(PartitionKey partitionKey, FileFormat format,
+                                  OutputFileFactory fileFactory, FileIO io,
+                                  long targetFileSize, ContentFileWriterFactory<ContentFileT, T> writerFactory) {
+    this.partitionKey = partitionKey;
+    this.format = format;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+    this.writerFactory = writerFactory;
+
+    this.resultBuilder = WriterResult.builder();
+
+    openCurrent();
+  }
+
+  public CharSequence currentPath() {
+    return currentFile.encryptingOutputFile().location();
+  }
+
+  public long currentPos() {
```

Comment:
`pos` is typically used to indicate position in bytes rather than rows. How about changing this to something like `rowCount` instead?